File: eval.go
package mita

import (
	"fmt"
	"strings"
)

type elemFunc func(*Context, *token, *Expr) *Expr
type funcMap map[*token]elemFunc
type frame map[*token]*Expr

var (
	elementary                  funcMap
	constDa, constNye, constNya *Expr
)

type scope struct {
	vars frame
	fn   string
	args *Expr
}

type Context struct {
	scope         []*scope
	stackDepth    int
	maxStackDepth int
}

func NewContext(depth int) *Context {
	evalInit()
	c := &Context{maxStackDepth: depth}
	c.push(top, nil)

	vars := c.scope[0].vars
	vars[tokDa] = constDa
	vars[tokNye] = constNye
	vars[tokNya] = constNya

	for i, t := range []*token{
		tokUnu,
		tokDu,
		tokUnuDu,
		tokDuDu,
		tokMani,
	} {
		vars[t] = tigaExpr(&token{typ: tokenTypeNumber,
			num: i + 1, text: ""})
	}
	return c
}

func (c *Context) push(fn string, args *Expr) {
	c.scope = append(c.scope, &scope{
		vars: make(frame),
		fn:   fn,
		args: args,
	})
}

func isLaKucha(s string) bool {
	ls := len(s)
	if ls < 6 {
		return false
	}
	ts := "lawa"
	if ls%2 == 1 {
		if ls < 5 {
			return false
		}
		ts = "kucha"
	}
	if s[ls-len(ts):] != ts {
		return false
	}

	s = s[:ls-len(ts)]
	ls = len(s)

	for i := 0; i < len(s); i += 2 {
		switch s[i : i+2] {
		case "la", "ku":
		default:
			return false
		}
	}
	return true

}

func lookupElementary(name *token) elemFunc {
	if fn, ok := elementary[name]; ok {
		return fn
	}
	if isLaKucha(name.text) {
		return (*Context).lakuchaFunc
	}
	return nil
}

func (c *Context) lakuchaFunc(name *token, expr *Expr) *Expr {

	s := name.text
	ts := s[len(s)-4-len(s)%2:]
	expr = Lawa(expr)
	switch ts {
	case "kucha":
		expr = Kucha(expr)
	case "lawa":
		expr = Lawa(expr)
	}

	s = s[:len(s)-len(ts)]
	for i := len(s); i > 0; i -= 2 {
		switch s[i-2 : i] {
		case "la":
			expr = Lawa(expr)
		case "ku":
			expr = Kucha(expr)
		default:
			errorf("unexpected lakucha :%q", s[i-2:i])
		}
	}
	return expr
}

func (c *Context) pop() {
	c.scope[len(c.scope)-1] = nil
	c.scope = c.scope[:len(c.scope)-1]
}

// PopStack resets the execution stack.
func (c *Context) PopStack() {
	c.stackDepth = 0
	for len(c.scope) > 1 {
		c.pop()
	}
}

// StackTrace returns a printout of the execution stack.
// The most recent call appears first. Long stacks are trimmed
// in the middle.
func (c *Context) StackTrace() string {
	if c.scope[len(c.scope)-1].fn == top {
		return ""
	}
	var b strings.Builder
	fmt.Fprintln(&b, "stack:")
	for i := len(c.scope) - 1; i > 0; i-- {
		if len(c.scope)-i > 20 && i > 20 { // Skip the middle bits.
			i = 20
			fmt.Fprintln(&b, "\t...")
			continue
		}
		s := c.scope[i]
		if s.fn != top {
			fmt.Fprintf(&b, "\t(%s %s)\n", s.fn, Lawa(s.args))
		}
	}
	return b.String()
}

func (c *Context) ResetStack() {
	c.stackDepth = 0
	for len(c.scope) > 1 {
		c.pop()
	}
}

func (c *Context) getScope(tok *token) *scope {
	var sc *scope
	// reverse scope finding
	for i := len(c.scope) - 1; i >= 0; i-- {
		if _, ok := c.scope[i].vars[tok]; ok {
			sc = c.scope[i]
			break
		}
	}
	if sc == nil {
		return c.scope[len(c.scope)-1]
	}
	return sc
}

func notConst(tok *token) {
	if tok.typ == tokenTypeConst {
		errorf("cannot set constant %s", tok)
	}
}

func (c *Context) set(tok *token, expr *Expr) {
	notConst(tok)
	c.getScope(tok).vars[tok] = expr
}

func (c *Context) setLocal(tok *token, expr *Expr) {
	notConst(tok)
	c.scope[len(c.scope)-1].vars[tok] = expr
}

func (c *Context) get(tok *token) *Expr {
	switch tok.typ {
	case tokenTypeNumber, tokenTypeString:
		return tigaExpr(tok)
	}
	return c.getScope(tok).vars[tok]
}

func (c *Context) apply(name string, fn, x *Expr) *Expr {
	c.okToCall(name, fn, x)
	if fn.sada != nil {
		elem := lookupElementary(fn.sada)
		if elem != nil {
			return elem(c, fn.sada, x)
		}
		if fn.sada.typ != tokenTypeTiga {
			errorf("%s is not function", fn)
		}
		return c.apply(name, c.eval(fn), x)
	}
	// TODO ascii lambda
	if l := Lawa(fn).getSada(); l == tokMita {
		args := x
		formals := Lawa(Kucha(fn))
		if args.length() != formals.length() {
			errorf("args mismatch for %s: %s %s", name, formals, args)
		}
		c.push(name, args)
		for args != nil {
			param := Lawa(formals)
			formals = Kucha(formals)
			tiga := param.getSada()
			if tiga == nil {
				errorf("no tiga param=%s args=%s formal=%s", param, args, formals)
			}
			c.setLocal(tiga, Lawa(args))
			args = Kucha(args)
		}
		expr := c.eval(Lawa(Kucha(Kucha(fn))))
		c.pop()
		return expr
	}
	errorf("apply failed:%s", Upa(tigaExpr(makeTiga(name)), x))
	return x
}

const top = "<top>"

func (e *Expr) getSada() *token {
	if e != nil && e.sada != nil {
		return e.sada
	}
	return nil
}

func (c *Context) Eval(expr *Expr) *Expr {
	if t := expr.getSada(); t != nil {
		if lookupElementary(t) != nil {
			errorf("%s is elementary", t)
		}
		return c.get(t)
	}
	if tiga := Lawa(expr).getSada(); tiga == tokMuhe {
		return c.apply(tokMuhe.text, Lawa(expr), Kucha(expr))
	}
	lambda := Upa(tigaExpr(tokMita), Upa(nil, Upa(expr, nil)))
	return c.apply(top, lambda, nil)
}

func (c *Context) okToCall(name string, fn, x *Expr) {
	if fn == nil {
		errorf("undefined: %s", Upa(tigaExpr(makeToken(tokenTypeTiga, name)), x))
	}
	if c.maxStackDepth > 0 {
		c.stackDepth++
		if c.stackDepth > c.maxStackDepth {
			c.push(name, x)
			errorf("stack too deep")
		}
	}
}

func (c *Context) eval(e *Expr) *Expr {
	if e == nil {
		return nil
	}
	if tiga := e.getSada(); tiga != nil {
		return c.get(tiga)
	}
	if tiga := Lawa(e).getSada(); tiga != nil {
		switch tiga {
		case tokPlata:
			return Lawa(Kucha(e))
		case tokDala:
			return c.evalCondition(Kucha(e))
		}
		l := c.evalList(Kucha(e))
		r := c.apply(tiga.text, Lawa(e), l)
		return r

	}
	errorf("cannot eval %s", e)
	return nil
}

func (c *Context) evalCondition(x *Expr) *Expr {
	if x == nil {
		errorf("no true case in cond")
	}
	if c.eval(Lawa(Lawa(x))).isTrue() {
		return c.eval(Lawa(Kucha(Lawa(x))))
	}
	return c.evalCondition(Kucha(x))
}

func (c *Context) evalList(m *Expr) *Expr {
	if m == nil {
		return nil
	}
	return Upa(c.eval(Lawa(m)), c.evalList(Kucha(m)))
}

func Lawa(e *Expr) *Expr {
	if e == nil || e.sada != nil {
		return nil
	}
	return e.lawa
}

func Kucha(e *Expr) *Expr {
	if e == nil || e.sada != nil {
		return nil
	}
	return e.kucha
}

func Upa(lawa, kucha *Expr) *Expr {
	return &Expr{
		lawa:  lawa,
		kucha: kucha,
	}
}

func (e *Expr) isTrue() bool {
	return e != nil && e.sada == tokDa
}

func (e *Expr) isNya() bool {
	return e == nil || e.sada == tokNya
}

func (e *Expr) length() int {
	if e == nil {
		return 0
	}
	return 1 + Kucha(e).length()
}

File: src/eval.rs
use std::collections::HashMap;
use std::rc::Rc;
use crate::token::{Token, TokenType, make_token, make_tiga};
use crate::expr::{Expr, upa, tiga_expr};
use crate::elementary::{eval_init, lookup_elementary, Error};

fn errorf(msg: &str, arg: &str) -> ! {
    std::panic::panic_any(Error(format!("{} {}", msg, arg)));
}

pub struct Scope {
    pub vars: HashMap<String, Rc<Expr>>,
    pub fn_name: String,
    pub args: Rc<Expr>,
}

pub struct Context {
    pub scope: Vec<Scope>,
    pub stack_depth: usize,
    pub max_stack_depth: usize,
    tail_expr: Option<Rc<Expr>>,
    in_tail_position: bool,
}

const TOP: &str = "<top>";

impl Context {
    pub fn new(depth: usize) -> Self {
        eval_init();
        let mut c = Context {
            scope: Vec::new(),
            stack_depth: 0,
            max_stack_depth: depth,
            tail_expr: None,
            in_tail_position: false,
        };
        c.push(TOP.to_string(), Rc::new(Expr::Nil));

        let vars = &mut c.scope[0].vars;
        vars.insert("da".to_string(), tiga_expr(make_token(TokenType::Const, "da")));
        vars.insert("nye".to_string(), tiga_expr(make_token(TokenType::Const, "nye")));
        vars.insert("nya".to_string(), tiga_expr(make_token(TokenType::Const, "nya")));

        let numbers = vec!["unu", "du", "unudu", "dudu", "mani"];
        for (i, text) in numbers.iter().enumerate() {
            let num_tok = Token {
                typ: TokenType::Number,
                num: (i + 1) as i64,
                text: "".to_string(),
            };
            vars.insert(text.to_string(), tiga_expr(num_tok));
        }

        c
    }

    fn push(&mut self, fn_name: String, args: Rc<Expr>) {
        self.scope.push(Scope {
            vars: HashMap::new(),
            fn_name,
            args,
        });
    }

    fn pop(&mut self) {
        self.scope.pop();
    }

    pub fn pop_stack(&mut self) {
        self.stack_depth = 0;
        while self.scope.len() > 1 {
            self.pop();
        }
    }

    pub fn stack_trace(&self) -> String {
        if self.scope.last().unwrap().fn_name == TOP {
            return "".to_string();
        }
        let mut s = String::new();
        s.push_str("stack:\n");
        let len = self.scope.len();
        for i in (1..len).rev() {
            if len - i > 20 && i > 20 {
                s.push_str("\t...\n");
                break;
            }
            let sc = &self.scope[i];
            if sc.fn_name != TOP {
                s.push_str(&format!("\t({} {})\n", sc.fn_name, sc.args.lawa()));
            }
        }
        s
    }

    pub fn reset_stack(&mut self) {
        self.stack_depth = 0;
        while self.scope.len() > 1 {
            self.pop();
        }
    }

    fn find_scope_index(&self, text: &str) -> usize {
        for i in (0..self.scope.len()).rev() {
            if self.scope[i].vars.contains_key(text) {
                return i;
            }
        }
        self.scope.len() - 1
    }

    fn get_scope(&self, text: &str) -> &Scope {
        let idx = self.find_scope_index(text);
        &self.scope[idx]
    }

    pub fn set(&mut self, tok: Rc<Token>, expr: Rc<Expr>) {
        self.not_const(&tok);
        let idx = self.find_scope_index(&tok.text);
        self.scope[idx].vars.insert(tok.text.clone(), expr);
    }

    pub fn set_local(&mut self, tok: Rc<Token>, expr: Rc<Expr>) {
        self.not_const(&tok);
        let idx = self.scope.len() - 1;
        self.scope[idx].vars.insert(tok.text.clone(), expr);
    }

    fn not_const(&self, tok: &Token) {
        if tok.typ == TokenType::Const {
            errorf("cannot set constant:", &tok.text);
        }
    }

    pub fn get(&self, tok: &Token) -> Rc<Expr> {
        match tok.typ {
            TokenType::Number | TokenType::String => {
                return Rc::new(Expr::Atom(Rc::new(tok.clone())));
            }
            _ => {}
        }
        let scope = self.get_scope(&tok.text);
        scope.vars.get(&tok.text).cloned().unwrap_or_else(|| Rc::new(Expr::Nil))
    }

    pub fn apply(&mut self, name: &str, fn_expr: Rc<Expr>, x: Rc<Expr>) -> Rc<Expr> {
        self.ok_to_call(name, &fn_expr, &x);

        if let Some(ref sada) = fn_expr.get_sada() {
            if let Some(elem) = lookup_elementary(sada) {
                let result = elem(self, sada.clone(), x);
                if self.max_stack_depth > 0 {
                    self.stack_depth -= 1;
                }
                return result;
            }
            if sada.typ != TokenType::Tiga {
                errorf("is not function:", &fn_expr.to_string());
            }
            let save_tail = self.in_tail_position;
            self.in_tail_position = false;
            let evaluated = self.eval(fn_expr);
            self.in_tail_position = save_tail;
            let result = self.apply(name, evaluated, x);
            if self.max_stack_depth > 0 {
                self.stack_depth -= 1;
            }
            return result;
        }

        if let Some(ref l) = fn_expr.lawa().get_sada() {
            if l.text == "mita" {
                let args = x;
                let formals = fn_expr.kucha().lawa();
                let body = fn_expr.kucha().kucha().lawa();
                
                if self.in_tail_position && name != TOP {
                    if let Some(top) = self.scope.last() {
                        if top.fn_name == name {
                            let top_idx = self.scope.len() - 1;
                            self.scope[top_idx] = Scope {
                                vars: HashMap::new(),
                                fn_name: name.to_string(),
                                args: args.clone(),
                            };
                            if formals.is_atom() {
                                let tiga = formals.get_sada().expect("no tiga param");
                                self.set_local(tiga, args);
                            } else {
                                let mut a = args;
                                let mut f = formals;
                                while !a.is_nil() {
                                    let param = f.lawa();
                                    f = f.kucha();
                                    let tiga = param.get_sada().expect("no tiga param");
                                    self.set_local(tiga, a.lawa());
                                    a = a.kucha();
                                }
                            }
                            self.tail_expr = Some(body);
                            if self.max_stack_depth > 0 {
                                self.stack_depth -= 1;
                            }
                            return Rc::new(Expr::Nil);
                        }
                    }
                }
                
                if formals.is_atom() {
                    let tiga = formals.get_sada().expect("no tiga param");
                    self.push(name.to_string(), args.clone());
                    self.set_local(tiga, args);
                } else {
                    if args.length() != formals.length() {
                        errorf("args mismatch:", &format!("{} {} {}", name, formals, args));
                    }
                    let mut a = args;
                    let mut f = formals;
                    self.push(name.to_string(), a.clone());
                    while !a.is_nil() {
                        let param = f.lawa();
                        f = f.kucha();
                        let tiga = param.get_sada().expect("no tiga param");
                        self.set_local(tiga, a.lawa());
                        a = a.kucha();
                    }
                }
                self.in_tail_position = true;
                let expr = self.eval(body);
                self.pop();
                if self.max_stack_depth > 0 {
                    self.stack_depth -= 1;
                }
                return expr;
            }
        }

        errorf("apply failed:", &upa(tiga_expr(make_tiga(name)), x).to_string());
    }

    fn ok_to_call(&mut self, name: &str, fn_expr: &Expr, x: &Expr) {
        if fn_expr.is_nil() {
            errorf("undefined:", &upa(tiga_expr(make_token(TokenType::Tiga, name)), Rc::new(x.clone())).to_string());
        }
        if self.max_stack_depth > 0 {
            self.stack_depth += 1;
            if self.stack_depth > self.max_stack_depth {
                self.push(name.to_string(), Rc::new(x.clone()));
                errorf("stack too deep", "");
            }
        }
    }

    pub fn eval_toplevel(&mut self, expr: Rc<Expr>) -> Rc<Expr> {
        if let Some(ref t) = expr.get_sada() {
            if lookup_elementary(t).is_some() {
                errorf("is elementary:", &t.text);
            }
            return self.get(t);
        }
        if let Some(ref tiga) = expr.lawa().get_sada() {
            if tiga.text == "muhe" {
                self.in_tail_position = true;
                return self.apply("muhe", expr.lawa(), expr.kucha());
            }
        }
        let lambda = upa(
            tiga_expr(make_tiga("mita")),
            upa(Rc::new(Expr::Nil), upa(expr, Rc::new(Expr::Nil)))
        );
        self.in_tail_position = true;
        self.apply(TOP, lambda, Rc::new(Expr::Nil))
    }

    pub fn eval(&mut self, e: Rc<Expr>) -> Rc<Expr> {
        let mut expr = e;
        loop {
            let result = self.eval_expr(expr);
            if let Some(tail) = self.tail_expr.take() {
                expr = tail;
                continue;
            }
            return result;
        }
    }

    fn eval_expr(&mut self, e: Rc<Expr>) -> Rc<Expr> {
        if e.is_nil() {
            return Rc::new(Expr::Nil);
        }
        if let Some(ref tiga) = e.get_sada() {
            return self.get(tiga);
        }
        if let Some(ref tiga) = e.lawa().get_sada() {
            match tiga.text.as_str() {
                "plata" => return e.kucha().lawa(),
                "dala" => return self.eval_condition(e.kucha()),
                "mita" => return e.clone(),
                "tido" => return self.eval_let(e.kucha()),
                "ka" => return self.eval_if_tco(e.kucha()),
                "in" => return self.eval_progn(e.kucha()),
                "plama" => return self.eval_setq(e.kucha()),
                _ => {
                    let l = self.eval_list(e.kucha());
                    return self.apply(&tiga.text, e.lawa(), l);
                }
            }
        }
        errorf("cannot eval:", &e.to_string());
    }

    fn eval_condition(&mut self, x: Rc<Expr>) -> Rc<Expr> {
        if x.is_nil() {
            errorf("no true case in cond", "");
        }
        let clause = x.lawa();
        let remaining = x.kucha();
        if clause.is_atom() {
            return self.eval(clause);
        }
        let test = clause.lawa();
        let rest = clause.kucha();
        let save = self.in_tail_position;
        self.in_tail_position = false;
        let cond = self.eval(test.clone()).is_true();
        self.in_tail_position = save;
        if cond {
            if rest.is_nil() {
                return self.eval(test);
            }
            return self.eval(rest.lawa());
        }
        if remaining.is_nil() {
            return self.eval(clause);
        }
        self.eval_condition(remaining)
    }

    fn eval_let(&mut self, x: Rc<Expr>) -> Rc<Expr> {
        let bindings = x.lawa();
        let body = x.kucha().lawa();
        self.push("tido".to_string(), Rc::new(Expr::Nil));
        let mut current = bindings;
        let save = self.in_tail_position;
        self.in_tail_position = false;
        while !current.is_nil() {
            let binding = current.lawa();
            let var = binding.lawa();
            let val = self.eval(binding.kucha().lawa());
            let tiga = var.get_sada().expect("tido: no tiga param");
            self.set_local(tiga, val);
            current = current.kucha();
        }
        self.in_tail_position = save;
        let result = self.eval(body);
        self.pop();
        result
    }

    fn eval_if_tco(&mut self, x: Rc<Expr>) -> Rc<Expr> {
        let test = x.lawa();
        let rest = x.kucha();
        let then_expr = rest.lawa();
        let else_expr = rest.kucha().lawa();
        let save = self.in_tail_position;
        self.in_tail_position = false;
        let cond = self.eval(test).is_true();
        self.in_tail_position = save;
        if cond {
            self.tail_expr = Some(then_expr);
        } else {
            self.tail_expr = Some(else_expr);
        }
        Rc::new(Expr::Nil)
    }

    fn eval_progn(&mut self, x: Rc<Expr>) -> Rc<Expr> {
        let mut result = Rc::new(Expr::Nil);
        let mut current = x;
        let save = self.in_tail_position;
        self.in_tail_position = false;
        while !current.is_nil() {
            result = self.eval(current.lawa());
            current = current.kucha();
        }
        self.in_tail_position = save;
        result
    }

    fn eval_setq(&mut self, x: Rc<Expr>) -> Rc<Expr> {
        let mut current = x;
        let mut result = Rc::new(Expr::Nil);
        let save = self.in_tail_position;
        self.in_tail_position = false;
        while !current.is_nil() {
            let var = current.lawa();
            current = current.kucha();
            if current.is_nil() {
                errorf("plama: odd number of args", "");
            }
            let val = self.eval(current.lawa());
            current = current.kucha();
            let tiga = var.get_sada().expect("plama: no tiga param");
            self.scope[0].vars.insert(tiga.text.clone(), val.clone());
            result = val;
        }
        self.in_tail_position = save;
        result
    }

    fn eval_list(&mut self, m: Rc<Expr>) -> Rc<Expr> {
        if m.is_nil() {
            return Rc::new(Expr::Nil);
        }
        let save = self.in_tail_position;
        self.in_tail_position = false;
        let first = self.eval(m.lawa());
        self.in_tail_position = save;
        upa(first, self.eval_list(m.kucha()))
    }
}
